api/server/router/system/system_routes.go

```diff
@@ -90,24 +90,45 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
 }
 
 func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+	if err := httputils.ParseForm(r); err != nil {
+		return err
+	}
+
+	var objectTypes map[types.DiskUsageObject]struct{}
+	if typeStrs, ok := r.Form["types"]; ok {
+		objectTypes = make(map[types.DiskUsageObject]struct{}, 4)
+		for _, ts := range typeStrs {
+			switch typ := types.DiskUsageObject(ts); typ {
+			case types.ContainerObject, types.ImageObject, types.VolumeObject, types.BuildCacheObject:
+				objectTypes[typ] = struct{}{}
+			default:
+				return invalidRequestError{fmt.Errorf("unknown object type: %s", typ)}
+			}
+		}
+	}
+
 	eg, ctx := errgroup.WithContext(ctx)
 
 	var du *types.DiskUsage
 	eg.Go(func() error {
 		var err error
-		du, err = s.backend.SystemDiskUsage(ctx)
+		du, err = s.backend.SystemDiskUsage(ctx, DiskUsageOptions{
+			ObjectTypes: objectTypes,
+		})
 		return err
 	})
 
 	var buildCache []*types.BuildCache
-	eg.Go(func() error {
-		var err error
-		buildCache, err = s.builder.DiskUsage(ctx)
-		if err != nil {
-			return pkgerrors.Wrap(err, "error getting build cache usage")
-		}
-		return nil
-	})
+	if _, ok := objectTypes[types.BuildCacheObject]; ok || len(objectTypes) == 0 {
```

Comment:
Just like in the daemon code for disk usage, maybe have a default value that adds all the object types so `len(objectTypes)` is never `0` (which wouldn't make much sense from an API perspective).

Reply:
The idea here was that the backend continues to have its own default (if no options passed, everything is returned).

It's a bit unfortunate that `build cache` is the odd one out here, and not provided by the daemon/backend (in which case this special handling wouldn't be needed here.

Wondering it it would make sense to integrate buildkit there instead of here (not sure)